pkg/apiserver/domain/service/pipeline.go

```diff
@@ -203,53 +320,544 @@ func (p pipelineRunServiceImpl) StopPipelineRun(ctx context.Context, pipelineRun
+	// valueFromStep know which step the value came from
+	valueFromStep := make(map[string]string)
+	inputsSpec := make(map[string]v1alpha1.StepInputs)
+	stepInputs := make([]apis.StepInputBase, 0)
+	if pipelineRun.Spec.WorkflowSpec != nil {
+		for _, step := range pipelineRun.Spec.WorkflowSpec.Steps {
+			if step.Inputs != nil {
+				inputsSpec[step.Name] = step.Inputs
+			}
+			if step.Outputs != nil {
+				for _, o := range step.Outputs {
+					valueFromStep[o.Name] = step.Name
+				}
+			}
+			for _, sub := range step.SubSteps {
+				if sub.Inputs != nil {
+					inputsSpec[sub.Name] = sub.Inputs
+				}
+				if sub.Outputs != nil {
+					for _, o := range sub.Outputs {
+						valueFromStep[o.Name] = sub.Name
+					}
+				}
+			}
+		}
+	}
+	ctxBackend := pipelineRun.Status.ContextBackend
+	if ctxBackend == nil {
+		return apis.GetPipelineRunInputResponse{}, nil
+	}
+	v, err := wfUtils.GetDataFromContext(ctx, p.KubeClient, ctxBackend.Name, pipelineRun.PipelineRunName, ctxBackend.Namespace)
+	if err != nil {
+		log.Logger.Errorf("get data from context backend failed: %v", err)
+		return apis.GetPipelineRunInputResponse{}, bcode.ErrGetContextBackendData
+	}
+	for _, s := range pipelineRun.Status.Steps {
+		var (
+			subStepStatus *v1alpha1.StepStatus
+			ok            bool
+		)
+		if stepName != "" && s.Name != stepName {
+			subStepStatus, ok = haveSubSteps(s, stepName)
+			if !ok {
+				continue
+			}
+			subVars := getStepInputs(*subStepStatus, inputsSpec, v, valueFromStep)
+			stepInputs = append(stepInputs, subVars)
+			break
+		}
+		stepInputs = append(stepInputs, getStepInputs(s.StepStatus, inputsSpec, v, valueFromStep))
+		for _, sub := range s.SubStepsStatus {
+			stepInputs = append(stepInputs, getStepInputs(sub, inputsSpec, v, valueFromStep))
+		}
+		if stepName != "" && s.Name == stepName {
+			// already found the step
+			break
+		}
+	}
+	return apis.GetPipelineRunInputResponse{StepInputs: stepInputs}, nil
+}
+
+func haveSubSteps(step v1alpha1.WorkflowStepStatus, subStep string) (*v1alpha1.StepStatus, bool) {
+	for _, s := range step.SubStepsStatus {
+		if s.Name == subStep {
+			return &s, true
+		}
+	}
+	return nil, false
+}
+
+func (p pipelineRunServiceImpl) GetPipelineRunLog(ctx context.Context, pipelineRun apis.PipelineRun, step string) (apis.GetPipelineRunLogResponse, error) {
+	project := ctx.Value(&apis.CtxKeyProject).(*model.Project)
+	if pipelineRun.Status.ContextBackend == nil {
+		return apis.GetPipelineRunLogResponse{}, nil
+	}
+
+	logConfig, err := wfUtils.GetLogConfigFromStep(ctx, p.KubeClient, pipelineRun.Status.ContextBackend.Name, pipelineRun.PipelineName, project.GetNamespace(), step)
+	if err != nil {
+		if strings.Contains(err.Error(), "no log config found") {
+			return apis.GetPipelineRunLogResponse{
+				StepBase: getStepBase(pipelineRun, step),
+				Log:      "",
+			}, nil
+		}
+		return apis.GetPipelineRunLogResponse{}, err
+	}
+	var logs string
+	switch {
+	case logConfig.Data:
```

Comment:
What means `logConfig.Data`?  Special implement for the job logs?

Reply:
See log config definition here: https://github.com/chivalryq/kubevela/blob/75760521a460d881d206739fb1c33ccd5884f50c/pkg/stdlib/pkgs/util.cue
https://github.com/kubevela/kubevela/blob/75760521a460d881d206739fb1c33ccd5884f50c/pkg/stdlib/pkgs/util.cue